backend/app/commands/CreateIngestion.scala

```diff
@@ -13,22 +13,47 @@ import scala.concurrent.ExecutionContext
 case class CreateIngestion(data: CreateIngestionRequest, collectionUri: Uri, manifest: Manifest, index: Index, pages: Pages)
                           (implicit ec: ExecutionContext) extends AttemptCommand[Uri] with Logging {
   def process(): Attempt[Uri] = {
+    val (ingestionId, ingestionUri) = provideIngestionUri()
 
-    val providedDisplayName = data.name.toAttempt {
-      manifest.getIngestionCount(collectionUri).map { i =>
-        s"Ingestion ${(i + 1).toString}"
-      }
+    createIngestion(ingestionId, ingestionUri)
+  }
+
+  def createOrGet(): Attempt[Uri] = {
+    val (ingestionId, ingestionUri) = provideIngestionUri()
+    val current = ingestionUri.flatMap(uri => manifest.getIngestion(uri))
+
+    current.map { c =>
+      Uri(c.uri)
+    }.recoverWith {
+      case f: NotFoundFailure =>
+        createIngestion(ingestionId, ingestionUri)
     }
+  }
 
-    val langs = data.languages.flatMap(Languages.getByKey)
+  private def provideIngestionUri(): (Attempt[String], Attempt[Uri]) = {
+    val providedDisplayName = getProvidedDisplayName()
 
+    val ingestionId = providedDisplayName.map(d => UriCleaner.clean(d))
+    val ingestionUri = ingestionId.map(id => collectionUri.chain(id))
+    (ingestionId, ingestionUri)
+  }
+
+  private def createIngestion(ingestionId: Attempt[String], ingestionUri: Attempt[Uri]): Attempt[Uri] = {
```

Comment:
I think in general it might be a good idea to remove the Attempt from stuff before you pass it into a function so that functions don't immediately have to deal with the fact something might be a failure (and to avoid having to construct an Attempt object when e.g. testing the function). I had a go at rewriting this file with that in mind - I think it is a slight improvement - see what you think https://github.com/guardian/giant/commit/aa42028cf1c112631a854cad65c4a1c58a066cf1

Reply:
This is a great suggestion, thanks for pointing it out and also for doing it in a branch. I made the changes :)